### DISCUSSION TITLE
How to use appender with list types in C api

### DISCUSSION THREAD

**Discussion by thijsmie** [duckdb/duckdb#3740] [Q&A] [answered]
Hi there!

I'm trying to map a generated C type to DuckDB for a tool. These types can contain any level of nesting with lists and arrays so they are potentially quite complex. I've managed to write the mapping for structures, however, I am unsure of how to use the `appender` when lists are involved. There doesn't seem to be an `append` or `resize` on `duckdb_list_vector_*` like-methods so when creating a datachunk I am unable to fill the list. If you could give a C example of how to use the appender for something like `CREATE TABLE dummy (sample STRUCT(field INTEGER[]));` I think I can work from there.

**Comment by Mytherin** [comment]
The C API currently appears to only support reading lists, not writing them. It will need to be extended to support this. Happy to review a PR!

**Reply by calvin2021y** [reply]
can you explain how to read the list, I can not find a way to read it.

**Comment by calvin2021y** [comment]
I can not append list from C, and duckdb_append_blob has zero bug at v0.4.  this prevent me use duckdb .

please consider release a new version with fix the duckdb_append_blob

**Comment by prashanthellina** [comment]
@Mytherin Is there any way to handle Lists, Arrays etc in the Appender API by using `duckdb_append_data_chunk`?

**Comment by litao3rd** [comment]
Hi, is there any way to handle Lists, Arrays etc in the Appender API now

**Reply by Mytherin** [reply, marked as answer]
Yes, you can use `duckdb_append_data_chunk` to append any data type

**Reply by litao3rd** [reply]
It seems this is a c style function family. Is there any c++ style like Appender interface. Thanks very much

**Reply by Mytherin** [reply]
The C++ appender also has the [AppendDataChunk](https://github.com/duckdb/duckdb/blob/cc2dc1d6d4b29b1b20c94e70061de2f441b794a1/src/include/duckdb/main/appender.hpp#L88) method.

In C++ you can use the [Value](https://github.com/duckdb/duckdb/blob/cc2dc1d6d4b29b1b20c94e70061de2f441b794a1/src/include/duckdb/common/types/value.hpp#L4) class to append data of any type - this will be much easier to write code for, but will be significantly slower than constructing the data chunks directly. I would recommend trying the `Value` method and only going for the DataChunk method if performance is a big consideration.

**Reply by litao3rd** [reply]
Great! I successfully created a table with a `VARCHAR[]` column and appended data to the table using a `DataChunk`. Thank you for your guidance.
